omod/src/test/java/org/openmrs/module/datafilter/web/DataFilterWebFilterTest.java

```diff
@@ -31,6 +31,7 @@ public void doFilter_shouldAlwaysResetTheFiltersOnTheCurrentThread() throws Exce
 		ThreadLocal areFiltersSet = new ThreadLocal();
 		areFiltersSet.set(true);
 		Whitebox.setInternalState(DataFilterSessionContext.class, "areFiltersSet", areFiltersSet);
+		// FIXE: No idea why the bellow assert is failing
 		assertNotNull(((ThreadLocal) Whitebox.getInternalState(DataFilterSessionContext.class, "areFiltersSet")).get());
```

Comment:
It means **areFiltersSet** is not getting set which means the logic that sets it is broken by the changes, so you probably want to fix it.

Reply:
Hi @wluyima, I'll need some more help with this. Not sure were to start looking.

Reply:
Try this,

- Temporarily comment out that offending line
- Add a break point in DataFilterSessionContext.reset method where it clears the thread local and inspect the state of the thread local before it's cleared.

Reply:
I believe something changed in powermock so that they don't allow setting the value of a final field, which makes sense if you think about it, so the value of the thread local is is null since that's what it got initialized to, I don't know why it was possible in older versions to change the value of final field.

Anyways, with that said I think you need to update the test by replacing the line(s) that sets a new value for the field in the test since it's final and instead just set the value of the thread local instance to true

Reply:
It means that failing assertion line needs to instead check that the value of the thread local is set to true.

Reply:
Alternatively, just add these annotations to the test class, that way the final field is not yet initialized until we do so in the test method  (I think this is an easier fix and more elegant)

```
@RunWith(PowerMockRunner.class)
@PrepareForTest(DataFilterSessionContext.class)
```

Reply:
Thanks @wluyima, the change of behaviour regarding the final field would explain the issue. I've updated the pull request with the annotations above and, it's now working as expected. Thanks for your analisis.